xds/internal/server/rds_handler.go

```diff
@@ -89,74 +93,93 @@ func (rh *rdsHandler) updateRouteNamesToWatch(routeNamesToWatch map[string]bool)
 			delete(rh.updates, routeName)
 		}
 	}
+}
 
-	// If the full list (determined by length) of updates are now successfully
-	// updated, the listener is ready to be updated.
-	if len(rh.updates) == len(rh.cancels) && len(routeNamesToWatch) != 0 {
-		drainAndPush(rh.updateChannel, rdsHandlerUpdate{updates: rh.updates})
-	}
+// determines if all dynamic RDS needed has received configuration or update.
+func (rh *rdsHandler) determineRDSReady() bool {
+	return len(rh.updates) == len(rh.cancels)
 }
 
-// handleRouteUpdate persists the route config for a given route name, and also
-// sends an update to the Listener Wrapper on an error received or if the rds
-// handler has a full collection of updates.
-func (rh *rdsHandler) handleRouteUpdate(routeName string, update xdsresource.RouteConfigUpdate) {
-	rh.mu.Lock()
-	defer rh.mu.Unlock()
-	rh.updates[routeName] = update
-
-	// If the full list (determined by length) of updates have successfully
-	// updated, the listener is ready to be updated.
-	if len(rh.updates) == len(rh.cancels) {
-		drainAndPush(rh.updateChannel, rdsHandlerUpdate{updates: rh.updates})
+func (rh *rdsHandler) handleRouteUpdate(routeName string, update rdsWatcherUpdate) {
+	rwu, ok := rh.updates[routeName]
+	if !ok {
+		rwu = rdsWatcherUpdate{}
```

Comment:
Hmm good point. This does wai right now (and is how I envisioned it), but I forget why I decided to persist around that old error, when the only way to trigger an error after a good update is resource not found, which would then overwrite the error anyway. Switching it to write a new update struct from each code path also wai. The error handling is correct though, because you want to keep the update around expect in the case where you get a resource not found, which is then when you start failing at l7 level.

Reply:
Actually, I played around with it, and it bloats the size if I do it your way. You need to read from the map to get the old update in the case where you don't want to clear it. I do see what you're saying about not clearing error, so I added the clear in the case where you get an update. Due to the update taking precedence (this error only matters if update is nil and error is set (see listener wrapper), this is logically a no-op, but it's fine to add. Thanks.